Add --expand-pointer-predicates to goto-instrument

karkhaz opened the pull request:
This patchset adds `__CPROVER_points_to_valid_memory`. It tells CBMC that an object of a particular size is pointed to by a pointer, and can be used in assertions and assumptions.

It also adds the `--expand-pointer-predicates` flag to goto-instrument, which expands the `__CPROVER_points_to_valid_memory` directive. This is in preparation for future commits that will introduce more pointer predicates having side effects, which will need to be expanded into multiple instructions.

Commit:
Adds --expand-pointer-predicates to goto-instrument

The --expand-pointer-predicates pass resolves pointer predicates
(util/pointer_predicates.{cpp, h}) which have side effects and
so require expanding into multiple instructions.
Currently, the only such predicate that needs to be expanded is
__CPROVER_points_to_valid_memory (name subject to change).

The __CPROVER_points_to_valid_memory predicates takes two parameters,
a pointer and a size. Semantically,
__CPROVER_points_to_valid_memory(p, size) should be true if and only
if p points to a memory object which is valid to read/write for at
least size bytes past p. When used in assertions, this will be checked
in a similar manner to how --pointer-check checks validity of a
dereference. When used in assumptions, this predicate creates (if none
exists already) an object for p to refer to, using
local_bitvector_analysis to make that determination, and then makes
assumptions (again corresponding to the assertions made by
--pointer-check) about that object.

Commit:
Adds a one-argument points_to_valid_memory

This commit adds support for a one-argument form of
__CPROVER_points_to_valid_memory, taking only a pointer, with no size.

If ptr has type T*, then __CPROVER_points_to_valid_memory(ptr) is
equivalent to __CPROVER_points_to_valid_memory(ptr, sizeof(T)).

Commit:
Adds tests for --expand-pointer-predicates

karkhaz committed:
Whitespace fix

karkhaz commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2109,6 +2111,65 @@ exprt c_typecheck_baset::do_special_functions(
+    if(expr.arguments().size() == 2)
+    {
+// TODO: We should add some way to enforce that the second argument
+// has a reasonable type (coercible to __CPROVER_size_t).
```
Any suggestions about this?

tautschnig replied:
How about testing whether it's an `ID_unsignedbv` or `ID_signedbv` - I think that should do the trick.

karkhaz replied:
Thank you---did this & added a test for it.

karkhaz committed:
Check 2nd arg to points_to_valid_memory is size_t

This commit ensures that the second argument to the
__CPROVER_points_to_valid_memory directive can be coerced to a size_t,
and adds a test to check for the failing case.

tautschnig commented:
@karkhaz This is still failing CI.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2ddb39ba4372c26d315030a2711f5f933a5178cf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104266680

karkhaz commented:
@tautschnig do you know why the AWS CodeBuild build is failing? I tried re-pushing, it doesn't help.

Also, if I want to diagnose this problem in the future, what should I do? I can't currently see the results of that build.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f53b62ba430b7fa877bea8639a3470f58d8225a3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104280806

tautschnig commented on regression/contracts/argument_sanity/test.desc:
```diff
@@ -0,0 +1,9 @@
+CORE
+main.c
+--apply-code-contracts
+^EXIT=1$
```
This is the cause of the Windows failure: `goto-cl` exits with exit code 64 on errors (cf. other intentionally-failing tests, `git grep -w 64 regression/ansi-c`).

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -34,6 +34,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "anonymous_member.h"
 #include "padding.h"
 
+bool is_lvalue(const exprt &expr);
```
This shouldn't be necessary, use `#include <util/expr_util.h>`

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2109,6 +2111,45 @@ exprt c_typecheck_baset::do_special_functions(
+    }
+
+    exprt same_object_expr;
+    if(expr.arguments().size() == 2)
```
This is established above, no need to check again.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2109,6 +2111,45 @@ exprt c_typecheck_baset::do_special_functions(
+    {
+      UNREACHABLE;
+    }
+    same_object_expr.add_source_location() = source_location;
```
Maybe `points_to_valid_memory` should actually take a `source_locationt` as an additional argument?

tautschnig commented on src/goto-instrument/expand_pointer_predicates.cpp:
```diff
@@ -0,0 +1,211 @@
+  new_tmp_symbol(const typet &type, const source_locationt &source_location);
+};
+
+bool is_lvalue(const exprt &expr);
```
This shouldn't be necessary, use `#include <util/expr_util.h>`

tautschnig commented on src/goto-instrument/expand_pointer_predicates.cpp:
```diff
@@ -0,0 +1,211 @@
+      else
+      {
+        continue;
+      }
```
The `else` case seems unnecessary.

tautschnig commented on src/goto-instrument/expand_pointer_predicates.cpp:
```diff
@@ -0,0 +1,211 @@
+        d->source_location = assume_expr.source_location();
+        const symbol_exprt obj =
+          new_tmp_symbol(object_type, d->source_location).symbol_expr();
+        d->code = code_declt(obj);
```
Please use the recently added API doing `assume_code.add(goto_programt::make_decl(...));` to combine most of the above into a single instruction.

tautschnig commented on src/goto-instrument/expand_pointer_predicates.cpp:
```diff
@@ -0,0 +1,211 @@
+        goto_programt::targett a = assume_code.add_instruction(ASSIGN);
+        a->source_location = assume_expr.source_location();
+        a->code = code_assignt(pointer_expr, rhs);
+        a->code.add_source_location() = assume_expr.source_location();
```
Use `assume_code.add(goto_programt::make_assign(...));`

tautschnig commented on src/goto-instrument/expand_pointer_predicates.cpp:
```diff
@@ -0,0 +1,211 @@
+{
+  return get_fresh_aux_symbol(
+    type,
+    id2string(source_location.get_function()),
```
If possible, please use the actual function identifier as found in the function map instead of this user-level name. This will require passing in such an additional `irep_idt` to `new_tmp_symbol`.

tautschnig commented on src/util/expr_util.cpp:
```diff
@@ -8,17 +8,102 @@ Author: Daniel Kroening, kroening@kroening.com
-#include "namespace.h"
-#include "arith_tools.h"
+#include <algorithm>
+#include <unordered_set>
```
Leave the STL includes where they were, separated by a blank line.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2131,10 +2131,10 @@ exprt c_typecheck_baset::do_special_functions(
+        expr.arguments()[1].type().id() != ID_signedbv)
       {
-        err_location(f_op);
+        error().source_location = f_op.source_location();
```
I'd prefer for those changes to be squashed into the earlier commit as otherwise that code wouldn't compile (which makes bisecting hard).

TGWDB commented:
Closing as this has been untouched for a couple of years. Please feel free to re-open/update if you believe this closure is erroneous.
